calculator.go

```diff
@@ -1,13 +1,111 @@
 // Package calculator provides a library for simple calculations in Go.
 package calculator
 
-// Add takes two numbers and returns the result of adding them together.
-func Add(a, b float64) float64 {
-return a + b
+import (
+	"errors"
+	"math"
+	"regexp"
+	"strconv"
+	"strings"
+)
+
+// Add takes 2+ numbers and returns the result of adding them together.
+func Add(a, b float64, rest ...float64) float64 {
+	result := a + b
+	for _, item := range rest {
+		result += item
+	}
+
+	return result
+}
+
+// Subtract takes 2+ numbers and returns the result of subtracting the second
+// from the first.  For variadic args, evaluation is (a - b) - c ...
+func Subtract(first float64, second float64, rest ...float64) float64 {
+	result := first - second
+
+	for _, item := range rest {
+		result -= item
+	}
+
+	return result
+
+}
+
+// Multiply multiplies 2+ numbers.  For variadic args, evaluation is (a * b) * c ...
+func Multiply(first float64, second float64, rest ...float64) float64 {
+	result := first * second
+
+	for _, item := range rest {
+		result *= item
+	}
+
+	return result
+
 }
 
-// Subtract takes two numbers and returns the result of subtracting the second
-// from the first.
-func Subtract(a, b float64) float64 {
-	return b - a
+// Divide divides 2+ numbers or returns an error if it would divide by 0.
+// For variadic args, evaluation is (a / b) / c ...
+func Divide(first float64, second float64, rest ...float64) (float64, error) {
+	if second == 0 {
+		return 0, errors.New("can't divide by zero")
+	}
+	result := first / second
+
+	for _, item := range rest {
+		if item == 0 {
+			return 0, errors.New("can't divide by zero")
+		}
+		result /= item
+	}
+
+	return result, nil
+}
+
+// Sqrt takes the square root of a nonnegative number or returns an error.
+func Sqrt(a float64) (float64, error) {
+	if a < 0 {
+		return 0, errors.New("can't take the square root of a negative number")
+	}
+	return math.Sqrt(a), nil
+}
+
+var validExprRe = regexp.MustCompile(`\d*(\.\d)?\s*[-+\/*]\s*\d*(\.\d)?$`)
+var operatorRe = regexp.MustCompile(`[-+\/*]`)
+
+// EvalExpr evaluates simple arithmetic expressions of "float64 operator float64"
+func EvalExpr(in string) (float64, error) {
+
+	if !validExprRe.MatchString(in) {
+		return 0, errors.New("could not parse input expression")
+	}
+
+	operatorMatch := operatorRe.FindAllString(in, -1)
```

Comment:
Have a look at `fmt.Sscanf` for parsing the expression!

Reply:
Won't this only handle things with spaces for the operator?  My implementation can handle `1+2`

Reply:
Yes, that's a good point; that's a rather irritating limitation, isn't it?

Reply:
Since a couple of other mentees have submitted working solutions for this using `fmt.Sscanf`, I've realised that both you and I misread the documentation here (or rather, it's misleading). Actually, it will parse expressions without spaces just fine (and it's very much to be preferred to regular expressions!).

Reply:
Why are regular expressions dispreferred, in general?

Reply:
I flailed against using Sscanf for an hour, and I am giving up.  I don't know why the code I have in right now would not match 1+2 correctly, but it doesn't.  My error is 
```
--- FAIL: TestEval (0.00s)
    /Users/bwarren/SideProjects/go-mentoring/ftl-fundamentals-bwarren2/calculator_test.go:207: unexpected error status: err Could not match the given operator: +2 given 1+2
FAIL
coverage: 82.9% of statements
FAIL	calculator	0.012s
FAIL```

I also have to swallow the EOF error, which feels unclean.  Why does this not work?

Reply:
I haven't dug into exactly why this is failing, but I suspect it's because you're scanning for `%v%v`, and `+2` is a valid number literal. So I think `Sscanf` is happily slurping `1` into the first variable, `+2` into the second, and then hitting EOF.

Here's a working format string that Bryan came up with: `%f%1s%f` (float, followed by length-1 string, followed by float)

We _don't_ want to swallow `io.EOF`, because it's telling us not enough values were matched by the scanning.

Reply:
> Why are regular expressions dispreferred?

Several reasons. One, they're slow as hell and expensive on memory. Two, they're hard to write, hard to understand, and hard to debug. Third, they're not Go. If we can solve the problem with Go, we prefer to do that, and only if there's something we can't do with Go code should we turn to other languages, such as regexps.

Reply:
The string length requirement was the ticket.  It is a bit weird that when I did specify the middle value as a string, but so it goes.